pkg/xds/translate.go

```diff
@@ -40,8 +41,186 @@ const (
 	HeaderScheme    = ":scheme"
 )
 
-// ClusterName specifies cluster name for a destination
-type ClusterName func(*routingv2.Destination) string
+const (
+	// UnresolvedCluster for destinations pointing to unknown clusters.
+	UnresolvedCluster = "unresolved-cluster"
+
+	// DefaultOperation is the default decorator
+	DefaultOperation = "default-operation"
+)
+
+// ServiceByName claims a service entry from the registry using a host name.
+type ServiceByName func(host string, contextNamespace string) *model.Service
+
+// SubsetSelector resolves a subset to labels.
+type SubsetSelector func(service *model.Service, subset string) map[string]string
+
+// GuardedHost is a context-dependent virtual host entry with guarded routes.
+type GuardedHost struct {
+	// Port is the capture port (e.g. service port)
+	Port int
+
+	// Services are the services matching the virtual host.
+	// The service host names need to be contextualized by the source.
+	Services []*model.Service
+
+	// Hosts is a list of alternative literal host names for the host.
+	Hosts []string
+
+	// Routes in the virtual host
+	Routes []GuardedRoute
+}
+
+// TranslateVirtualHosts creates the entire routing table for Istio v1alpha2 configs.
+// Services are indexed by FQDN hostnames.
+// Cluster domain is used to resolve short service names (e.g. "svc.cluster.local").
+func TranslateVirtualHosts(
+	serviceConfigs []model.Config,
+	services map[string]*model.Service,
+	subsetSelector SubsetSelector,
+	clusterDomain string) []GuardedHost {
+	out := make([]GuardedHost, 0)
+
+	serviceByName := func(host string, contextNamespace string) *model.Service {
+		if strings.Contains(host, ".") {
+			return services[host]
+		}
+
+		return services[fmt.Sprintf("%s.%s.%s", host, contextNamespace, clusterDomain)]
```

Comment:
What about where `contextNamespace` is empty?

Reply:
Can be ignore for now, but something to be careful about given the discussion on supporting Consul, CF etc

Reply:
Yes, definitely can be improved with more registry involvement. The current interface just uses FQDN, so not much we can do per registry. It would be nice to have per registry domain, for example.